CMakeLists.txt

```diff
@@ -84,8 +84,16 @@ if(CMAKE_COMPILER_IS_CLANG)
     set(CMAKE_C_FLAGS_RELEASE     "-O2")
     set(CMAKE_C_FLAGS_DEBUG       "-O0 -g3")
     set(CMAKE_C_FLAGS_COVERAGE    "-O0 -g3 --coverage")
-    set(CMAKE_C_FLAGS_ASAN        "-Werror -fsanitize=address -fno-common -fsanitize=undefined -fno-sanitize-recover -O3")
-    set(CMAKE_C_FLAGS_ASANDBG     "-Werror -fsanitize=address -fno-common -fsanitize=undefined -fno-sanitize-recover -O1 -g3 -fno-omit-frame-pointer -fno-optimize-sibling-calls ")
+    EXECUTE_PROCESS( COMMAND ${CMAKE_C_COMPILER} --version OUTPUT_VARIABLE clang_full_version_string )
+    string (REGEX REPLACE ".*clang version ([0-9]+\\.[0-9]+).*" "\\1" CLANG_VERSION_STRING ${clang_full_version_string})
```

Comment:
I have not tried this myself, but I think this should work on a Linux machine. However, I do not think this is a robust enough approach for the mbed TLS build system which is very portable. For example, I tried `clang --version` in mac and I got the following output:
```
Apple LLVM version 7.0.0 (clang-700.0.72)
Target: x86_64-apple-darwin14.4.0
Thread model: posix
```
In this case (and possibly many others), I do not think the regular expression will work. However the problem still needs to be solved, so I suggest:
1. Define which compiler versions we actually support. I think it would be good to involve @sbutcher-arm on this
2. Find an alternative mechanism to find out the compiler and version.
2.1 [CMAKE_C_COMPILER_ID and CMAKE_C_COMPILER_VERSION](https://cmake.org/cmake/help/v3.9/manual/cmake-toolchains.7.html?highlight=cmake_c_compiler#variables-and-properties) might be of help, though not sure how reliable that is.
2.2 Perhaps you could use a [try_compile()](https://cmake.org/cmake/help/v3.0/command/try_compile.html) to see what asan options the compiler accepts.
2.3 You could remove the Asan flags altogether and just document somewhere how to configure Cmake to build using Asan. We already need to configure this through the commandline for Makefile, but there are no instructions.

Reply:
I tried using `CMAKE_C_COMPILER_VERSION` but it did not work (returned an empty string)
If we want to remove Asan flags, it requires some further reasearch, more than just this simple bug fix

Reply:
@RonEld @andresag01 Similar to the current code, one could also execute `clang --help` and grep for `-fno-sanitize-recover=<value>` (as opposed to `-fno-sanitize-recover` at the end of a line). Not particularly elegant, though.

Reply:
@RonEld, @hanno-arm: The CMAKE documentation does say that CMAKE_C_COMPILER_ID and CMAKE_C_COMPILER version might not be set, so I suppose that is not an option either. However, I do think that the current solution is slightly fragile. I personally prefer just using `try_compile()` so that you run the compilation, if it fails, then you try changing the options and then hopefully it works... Not very elegant, but it might just work. In the case that the recompilation does not work then you would probably get a slightly confussing error message, but ...

Reply:
@andresag01 IMO it would be acceptable to document (and maybe even enforce that by appropriate directives in the CMakeLists file) that some cmake targets only work on Linux. For example [Clang's MemorySanitizer](https://clang.llvm.org/docs/MemorySanitizer.html) currently only works on Linux 64-bit.

Regardless of their current platform support in Clang (eg ASan is now multiplatform) I think we don't need to make the sanitizer targets as portable as possible: they're mainly a tool for us, not (IMO) something we owe to our users.